src/plugins/ui_stateful/clipboard.ts

```diff
@@ -55,18 +60,32 @@ export class ClipboardPlugin extends UIPlugin {
     switch (cmd.type) {
       case "CUT":
         const zones = this.getters.getSelectedZones();
-        const state = this.getClipboardState(zones, cmd.type);
-        return state.isCutAllowed(zones);
+        return this.isCutAllowedOnAnyPlugin(zones);
+      //@ts-ignore
+      case "PASTE_FROM_OS_CLIPBOARD":
+        this.copiedData = this.convertOSClipboardData(cmd.text);
       case "PASTE":
-        if (!this.state) {
+        if (!this.copiedData) {
           return CommandResult.EmptyClipboard;
         }
         const pasteOption =
           cmd.pasteOption || (this.paintFormatStatus !== "inactive" ? "onlyFormat" : undefined);
-        return this.state.isPasteAllowed(cmd.target, { pasteOption });
-      case "PASTE_FROM_OS_CLIPBOARD": {
-        const state = new ClipboardOsState(cmd.text, this.getters, this.dispatch, this.selection);
-        return state.isPasteAllowed(cmd.target, { pasteOption: cmd.pasteOption });
+        return this.isPasteAllowedOnAnyPlugin(cmd.target, this.copiedData, { pasteOption });
```

Comment:
handler, not plugin :) Actually I think you can juste change your method names to `this.isPasteAllowed`, `this.copy`, `this.paste` etc...